content/blog/2020-11-29-tracemonkey/index.md

```diff
@@ -0,0 +1,402 @@
+
+When the trace of a loop body is completed and the control flow
+returns to the start of the loop, if the present trace map matches the
+trace map used to start the trace (i.e. the loop is _type stable_),
+then the compiled trace can simply jump back to the start of the loop,
+"tying the knot". The loop will not run forever unless the loop in the
+original program ran forever because it will eventually cause a guard
+to fail, breaking out of native execution and following a different
+path of control flow.
+
+Not all loops are type stable. Consider the following (admittedly
+contrived) piece of code:
+
+```js
+var x;
+for (var i = 0; i < 100; ++i) {
+  x = 1;
+}
+```
+
+The first time the loop is encountered, the type of `x` is undefined,
+but the second time through it is now a number and remains that way
+for the rest of execution. Thus this loop converges to type stability.
+
+In such a situation, TraceMonkey will produce a seprate compiled trace
+for each type map, but eventually it will be able to tie the knot as
+the input and output types converge. Other more complicated
+convergence patterns are possible, including topologies where the
+control oscillates between two or more compiled traces, but these
+scenarious are still handled elegantly by this approach.
+
+### Integer specialization
+
+JavaScript does not have a native integer type; instead, all numbers
+are implemented as 64-bit floating point numbers,
+i.e. "doubles". However, many use cases of numbers use them only as
+integers, and processor integer instructions are considered cheaper
+than their floating-point counterparts, so it is desirable to use
+integer operations when possible.
+
+To this end, TraceMonkey speculatively assumes that all numbers are
+integers and backtracks if it is discovered that a particular variable
+must instead be a float. When such backtracking occurs, TraceMonkey
+records that that variable must be a float for the given trace in an
+"oracle". When restarting the loop, the trace map is constructed in
+part by referencing the oracle and not speculatively assigning any
+variable to be an integer if the oracle says otherwise.
+
+### Optimizations
+
+After the trace is recorded as LIR, nanojit runs a simple set of
+optimization passes followed by register allocation to produce machine
+code. As the LIR is a linear sequence of instructions in SSA form, all
+optimizations can be done in linear time and are straightforwardly
+local.
+
+The optimization passes are noticeably simpler than those used by
+static compilers because it is important not to spend too much time
+compiling while the program is executing. Other more recent JavaScript
+engines, including V8, go back and re-compile the same code with more
+sophisticated optimizations if it stays hot. TraceMonkey does not do
+this; perhaps such things are future improvements, or perhaps it was
+preferred to focus on the optimizations that deliver the best bang for
+your buck.
```

Comment:
Probably worth remembering the context here, i.e., TraceMonkey seems to have been the world's first JavaScript JIT. 2008 was a wild time of picking all the low-hanging fruit; more advanced techniques came later. Even Mozilla's current JS JIT uses [multiple tiers](https://hacks.mozilla.org/2019/08/the-baseline-interpreter-a-faster-js-interpreter-in-firefox-70/).